qtrees/data_processor.py

```diff
@@ -0,0 +1,311 @@
+            water = reduce(lambda left, right: pd.merge(left, right, on=["tree_id", "timestamp"],
+                                                        how='left'), [water, water_sga, water_gdk])
+            water = water.drop(columns="temp")
+            water = water.fillna(0)
+            water = water.groupby(["tree_id", "timestamp"]).rolling(window=ROLLING_WINDOW, min_periods=1).sum()
+            return water
+
+        def get_shading_index(relevant_trees):
+            monthly_shading = pd.read_sql(f"SELECT * FROM public.shading_monthly WHERE tree_id in {relevant_trees}", self.engine.connect())
+            shading_long = pd.melt(monthly_shading, id_vars="tree_id")
+            month_mapping = dict((v, k) for v, k in zip(shading_long.variable.unique(), range(1, 13)))
+            shading_long = shading_long.assign(month=[month_mapping[el] for el in shading_long.variable])
+            shading_long.drop(columns="variable", inplace=True)
+            shading_long.rename(columns={"value": "shading_index"}, inplace=True)
+            return shading_long
+
+        if self.with_sensors:
+            trees = get_sensors(subset)
+            if trees.empty:
+                return trees
+        else:
+            trees = subset
+            if self.date is None:
+                dates = pd.date_range("2021-06-01", pd.Timestamp("today") + pd.Timedelta(days=FORECAST_HORIZON if self.forecast else 0), tz="UTC")
+                dates = dates[dates.month.isin(range(4, 11))].to_series(name="timestamp")
+                trees = trees.merge(dates, how="cross")
+            else:
+                trees = trees.assign(timestamp=self.date)
+                trees = trees.astype({"timestamp": 'datetime64[ns, UTC]'})
+            trees = trees.assign(month=trees.timestamp.dt.month)
+        relevant_trees = tuple(trees.tree_id.unique())
+        if not self.public_run:
+            trees_private = pd.read_sql("SELECT tree_id,baumscheibe_m2,baumscheibe_surface FROM private.trees_private", self.engine.connect())
+            trees_private["baumscheibe_m2"] = pd.cut(trees_private["baumscheibe_m2"], bins=[0, 5, 100], labels=["klein", "groß"])
+            trees = trees.merge(trees_private, how="left", on="tree_id")
+            if not self.forecast:
+                water = get_watering(relevant_trees)
+                trees = trees.merge(water, how="left", on=["tree_id", "timestamp"])
+        shading = get_shading_index(relevant_trees)
+        trees = trees.merge(shading, how="left", on=["tree_id", "month"])
+        return trees
+
+    def _get_weather_measurements(self):
+        '''Get station data and solar if not public run and merge them. Generate weekly avg or sum depending on column.'''
+        if self.forecast and not self.public_run:
+            weather_station = pd.read_sql("SELECT tile_id, date, wind_max_ms, wind_avg_ms, temp_max_c, temp_avg_c, rainfall_mm FROM private.weather_tile_measurement", con=self.engine.connect())
```

Comment:
die liste wird hier paar mal genutzt. vielleicht auch wie unten in eine variable  weather_features rausziehen für bessere Wartbarkeit wenn was dazu kommt.